tools/schemapi/schemapi.py

```diff
@@ -1,89 +1,126 @@
-from itertools import chain, zip_longest
+from itertools import chain, groupby, islice, zip_longest
 from math import ceil
 from typing import (
     TYPE_CHECKING,
     Any,
     Dict,
-    Final,
     Iterable,
-    Iterator,
     List,
-    Literal,
+    Mapping,
     Sequence,
     TypeVar,
     Union,
     cast,
     overload,
 )
-from typing_extensions import TypeAlias
 
 import jsonschema
-import jsonschema.exceptions
 import jsonschema.validators
 import narwhals.stable.v1 as nw
+from jsonschema import ValidationError
 from packaging.version import Version
 
-# This leads to circular imports with the vegalite module. Currently, this works
-# but be aware that when you access it in this script, the vegalite module might
-# not yet be fully instantiated in case your code is being executed during import time
-from altair import vegalite
-
 if TYPE_CHECKING:
     from types import ModuleType
-    from typing import ClassVar
+    from typing import Callable, ClassVar, Final, Iterator, KeysView, Literal
 
-    from referencing import Registry
+    from jsonschema.protocols import Validator, _JsonParameter
 
     from altair.typing import ChartType
+    from altair.vegalite.v5.schema._typing import Map
 
     if sys.version_info >= (3, 13):
         from typing import TypeIs
     else:
         from typing_extensions import TypeIs
 
     if sys.version_info >= (3, 11):
-        from typing import Never, Self
+        from typing import LiteralString, Never, Self
+    else:
+        from typing_extensions import LiteralString, Never, Self
+    if sys.version_info >= (3, 10):
+        from typing import TypeAlias
     else:
-        from typing_extensions import Never, Self
+        from typing_extensions import TypeAlias
     _OptionalModule: TypeAlias = "ModuleType | None"
+    _Errs: TypeAlias = Iterable[ValidationError]
+    _ErrsLazy: TypeAlias = Iterator[ValidationError]
+    _ErrsLazyGroup: TypeAlias = Iterator[_ErrsLazy]
+    _IntoLazyGroup: TypeAlias = Iterator["tuple[str, ValidationError]"]
+    _ValidatorKeyword: TypeAlias = Literal[
+        "additionalProperties",
+        "enum",
+        "type",
+        "required",
+        "properties",
+        "anyOf",
+        "allOf",
+        "oneOf",
+        "ref",
+        "const",
+    ]
+    """Non-exhaustive listing of possible literals in ``ValidationError.validator``"""
+
+__all__ = [
+    "Optional",  # altair.utils
+    "SchemaBase",  # altair.vegalite.v5.schema.core
+    "Undefined",  # altair.utils
+    "UndefinedType",  # altair.vegalite.v5.schema.core -> (side-effect relied on to propagate to alt.__init__)
+    "_is_valid",  # altair.vegalite.v5.api
+    "_resolve_references",  # tools.schemapi.utils -> tools.generate_schema_wrapper
+    "_subclasses",  # altair.vegalite.v5.schema.core
+    "is_undefined",  # altair.typing
+    "validate_jsonschema",  # altair.utils.display
+    "with_property_setters",  # altair.vegalite.v5.schema.channels
+]
```

Comment:
Keeping track of the refs here, since you can't follow the references from the `tools` copy